### PR TITLE
Add issue/PR templates, code of conduct & better contributing guide

### PR DESCRIPTION
**Pull request by lewtun**

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,53 +1,259 @@
...
+
+   Whatever it is, we'd love to hear about it!
+
+2. Describe your requested feature in as much detail as possible. The more you can tell us about it, the better we'll be able to help you.
+3. Provide a *code snippet* that demonstrates the features usage.
+4. If the feature is related to a paper, please include a link.
+
+If your issue is well written we're already 80% of the way there by the time you create it.
+
+## Do you want to implement a new trainer?
```
Comment (lewtun) [inline CONTRIBUTING.md:83]: cc @lvwerra for viz on what we discussed offline

File: .github/workflows/ISSUE_TEMPLATE/bug-report.yml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+        Before you submit your bug report:
+
+          - If it is your first time submitting, be sure to check our [bug report guidelines](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#did-you-find-a-bug)
+
+  - type: textarea
+    id: system-info
+    attributes:
+      label: System Info
+      description: Please share your system info with us. You can run the command `transformers-cli env` and copy-paste its output below.
```
Comment (qgallouedec) [inline .github/workflows/ISSUE_TEMPLATE/bug-report.yml:18]: we should probably overload the cli so that it gets the trl version

Comment (lewtun) [reply .github/workflows/ISSUE_TEMPLATE/bug-report.yml:18]: Sure or we enable `trl env` to the existing CLI - I can do that in a follow-up PR

Comment (qgallouedec) [reply .github/workflows/ISSUE_TEMPLATE/bug-report.yml:18]: let's do this in a dedicated PR!

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+<!-- Remove if not applicable -->
+
+Fixes # (issue)
+
+
+## Before submitting
+- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
+- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
+      Pull Request section?
+- [ ] Was this discussed/approved via a Github issue? Please add a link
```
Comment (qgallouedec) [inline .github/PULL_REQUEST_TEMPLATE.md:22]:
```suggestion
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,53 +1,259 @@
...
+already reported** (use the search bar on GitHub under Issues). Your issue should also be related to bugs in the library itself, and not your code.
 
-First you want to make sure that all the tests pass:
+Once you've confirmed the bug hasn't already been reported, please include the following information in your issue so we can quickly resolve it:
 
-```bash
-make test
-```
+* Your **OS type and version** and **Python**, **PyTorch** and
+  **Transformers** versions when applicable.
```
Comment (qgallouedec) [inline CONTRIBUTING.md:57]:
```suggestion
* Your **OS type and version**, **Python**, **PyTorch**, **TRL** and **Transformers** versions.
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,53 +1,259 @@
...
+
+Based on the community and maintainer feedback, the next step will be to implement the trainer and config classes. See the following examples for inspiration:
+
+* Paired preference optimisation: [`dpo_trainer.py`](./trl/trainer/dpo_trainer.py) and [`dpo_config.py`](./trl/trainer/dpo_config.py)
+* RL-based optimisation: [`rloo_trainer.py](./trl/trainer/rloo_trainer.py) and [`rloo_config.py](./trl/trainer/rloo_config.py)
+* Online optimisation: [`online_dpo_trainer.py`](./trl/trainer/online_dpo_trainer.py) and [`online_dpo_config.py`](./trl/trainer/online_dpo_config.py)
+
+## Do you want to add documentation?
+
+We're always looking for improvements to the documentation that make it more clear and accurate. Please let us know how the documentation can be improved such as typos and any content that is missing, unclear or inaccurate. We'll be happy to make the changes or help you make a contribution if you're interested!
```
Comment (qgallouedec) [inline CONTRIBUTING.md:106]:
```suggestion
We're always looking for improvements to the documentation that make it more clear and accurate. Please let us know how the documentation can be improved, such as typos, dead links and any missing, unclear or inaccurate content.. We'll be happy to make the changes or help you make a contribution if you're interested!
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,53 +1,259 @@
...
+   **Do not** work on the `main` branch.
+
+4. Set up a development environment by running the following command in a conda or a virtual environment you've created for working on this library:
+
+   ```bash
+   $ make dev
+   ```
+
+   (If TRL was already installed in the virtual environment, remove
+   it with `pip uninstall tr;` before reinstalling it.)
```
Comment (qgallouedec) [inline CONTRIBUTING.md:160]:
```suggestion
   it with `pip uninstall trl` before reinstalling it.)
```

**Review by qgallouedec — APPROVED** [review]

**Pull request merged**